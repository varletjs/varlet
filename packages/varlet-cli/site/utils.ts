import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
