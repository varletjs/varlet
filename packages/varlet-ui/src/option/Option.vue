<template>
  <div
    class="var-option var--box"
    :class="[optionSelected ? 'var-option--selected-color' : null]"
    :style="{
      width: wrapWidth,
      color: optionSelected ? focusColor : undefined,
    }"
    v-ripple
    @click="handleClick"
  >
    <div
      class="var-option__cover"
      :class="[optionSelected ? 'var-option--selected-background' : null]"
      :style="{
        background: optionSelected ? focusColor : undefined,
      }"
    ></div>
    <var-checkbox
      ref="checkbox"
      :checked-color="focusColor"
      v-if="multiple"
      v-model="optionSelected"
      @click.stop
      @change="handleSelect"
    />

    <div class="var-option__text var--ellipsis">
      <slot>
        {{ label }}
      </slot>
    </div>
  </div>
</template>

<script lang="ts">
import VarCheckbox from '../checkbox'
import Ripple from '../ripple'
import { defineComponent, computed, ref, watch } from 'vue'
import { useSelect } from './provide'
import { props } from './props'
import type { Ref, ComputedRef } from 'vue'
import type { OptionProvider } from './provide'

export default defineComponent({
  name: 'VarOption',
  directives: { Ripple },
  components: {
    VarCheckbox,
  },
  props,
  setup(props) {
    const optionSelected: Ref<boolean> = ref(false)
    const selected: ComputedRef<boolean> = computed(() => optionSelected.value)
    const label: ComputedRef = computed(() => props.label)
    const value: ComputedRef = computed(() => props.value)
    const { select, bindSelect } = useSelect()
    const { wrapWidth, multiple, focusColor, onSelect } = select

    const handleClick = () => {
      optionSelected.value = !optionSelected.value
      onSelect(optionProvider)
    }

    const handleSelect = () => onSelect(optionProvider)

    const sync = (checked: boolean) => {
      optionSelected.value = checked
    }

    const optionProvider: OptionProvider = {
      label,
      value,
      selected,
      sync,
    }

    watch(
      [() => props.label, () => props.value],
      () => {
        if (props.label == null && props.value == null) {
          throw Error("Props label and value can't both be undefined\n")
        }
      },
      { immediate: true }
    )

    bindSelect(optionProvider)

    return {
      optionSelected,
      wrapWidth,
      multiple,
      focusColor,
      handleClick,
      handleSelect,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../ripple/ripple';
@import '../checkbox/checkbox';
@import '../select/select';
@import './option';
</style>
