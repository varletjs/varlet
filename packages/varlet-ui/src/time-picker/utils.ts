import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
