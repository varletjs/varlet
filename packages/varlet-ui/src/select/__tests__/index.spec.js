import example from '../example'
import Select from '..'
import VarSelect from '../Select'
import Option from '../../option'
import VarOption from '../../option/Option'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, trigger } from '../../utils/test'

test('test select example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test select plugin', () => {
  const app = createApp({}).use(Select)
  expect(app.component(Select.name)).toBeTruthy()
})

test('test option plugin', () => {
  const app = createApp({}).use(Option)
  expect(app.component(Option.name)).toBeTruthy()
})

const Wrapper = {
  components: {
    [VarSelect.name]: VarSelect,
    [VarOption.name]: VarOption,
  },
}

test('test select by label', async () => {
  const wrapper = mount({
    ...Wrapper,
    data: () => ({
      value: '',
    }),
    template: `
      <var-select v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
  })

  await wrapper.trigger('click')
  await trigger(document.querySelector('.var-option'), 'click')
  expect(wrapper.vm.value).toBe('吃饭')

  wrapper.unmount()
})

test('test select by value', async () => {
  const wrapper = mount({
    ...Wrapper,
    data: () => ({
      value: '',
    }),
    template: `
      <var-select v-model="value">
        <var-option label="吃饭" :value="1" />
        <var-option label="睡觉" :value="2" />
      </var-select>
    `,
  })

  await wrapper.trigger('click')
  await trigger(document.querySelector('.var-option'), 'click')
  expect(wrapper.vm.value).toBe(1)

  wrapper.unmount()
})

test('test select hint to be false', () => {
  const wrapper = mount({
    ...Wrapper,
    data: () => ({
      value: '吃饭',
    }),
    template: `
      <var-select :hint="false" v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
  })

  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
})

test('test select onFocus & onBlur', async () => {
  const onFocus = vi.fn()
  const onBlur = vi.fn()

  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: '',
      }),
      methods: {
        onFocus,
        onBlur,
      },
      template: `
      <div class="container"></div>
      <var-select v-model="value" @focus="onFocus" @blur="onBlur">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-select__menu').trigger('click')
  expect(onFocus).toHaveBeenCalledTimes(1)
  await wrapper.find('.container').trigger('click')
  expect(onBlur).toHaveBeenCalledTimes(1)

  wrapper.unmount()
})

test('test select disabled', async () => {
  const onFocus = vi.fn()
  const onBlur = vi.fn()

  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        disabled: true,
        value: '睡觉',
      }),
      methods: {
        onFocus,
        onBlur,
      },
      template: `
      <div class="container"></div>
      <var-select
        clearable
        :disabled="disabled"
        v-model="value"
        @focus="onFocus"
        @blur="onBlur"
      >
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-select__wrap').trigger('click')
  expect(onFocus).toHaveBeenCalledTimes(0)

  await wrapper.setData({ disabled: false })
  await wrapper.find('.var-select__wrap').trigger('click')
  await wrapper.setData({ disabled: true })
  await wrapper.find('.container').trigger('click')
  expect(onBlur).toHaveBeenCalledTimes(0)

  await wrapper.setData({ disabled: false })
  await wrapper.find('.var-select__wrap').trigger('click')
  await wrapper.setData({ disabled: true })
  await trigger(document.querySelector('.var-option'), 'click')
  expect(wrapper.vm.value).toBe('睡觉')

  await wrapper.find('.var-icon-close-circle').trigger('click')
  expect(wrapper.vm.value).toBe('睡觉')

  wrapper.unmount()
})

test('test select readonly', async () => {
  const onFocus = vi.fn()
  const onBlur = vi.fn()

  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        readonly: true,
        value: '睡觉',
      }),
      methods: {
        onFocus,
        onBlur,
      },
      template: `
      <div class="container"></div>
      <var-select
        clearable
        :readonly="readonly"
        v-model="value"
        @focus="onFocus"
        @blur="onBlur"
      >
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-select__wrap').trigger('click')
  expect(onFocus).toHaveBeenCalledTimes(0)

  await wrapper.setData({ readonly: false })
  await wrapper.find('.var-select__wrap').trigger('click')
  await wrapper.setData({ readonly: true })
  await wrapper.find('.container').trigger('click')
  expect(onBlur).toHaveBeenCalledTimes(0)

  await wrapper.setData({ readonly: false })
  await wrapper.find('.var-select__wrap').trigger('click')
  await wrapper.setData({ readonly: true })
  await trigger(document.querySelector('.var-option'), 'click')
  expect(wrapper.vm.value).toBe('睡觉')

  await wrapper.find('.var-icon-close-circle').trigger('click')
  expect(wrapper.vm.value).toBe('睡觉')

  wrapper.unmount()
})

test('test select clear', async () => {
  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: '吃饭',
      }),
      template: `
      <var-select clearable v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-icon-close-circle').trigger('click')
  expect(wrapper.vm.value).toBe(undefined)

  wrapper.unmount()
})

test('test select multiple value', async () => {
  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: [],
      }),
      template: `
      <var-select multiple v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-select__wrap').trigger('click')

  Array.from(document.querySelectorAll('.var-option')).forEach((el) => trigger(el, 'click'))
  expect(wrapper.vm.value).toStrictEqual(['吃饭', '睡觉'])

  wrapper.unmount()
})

test('test select multiple value in chips', async () => {
  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: [],
      }),
      template: `
      <var-select multiple chip v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  await wrapper.find('.var-select__wrap').trigger('click')

  Array.from(document.querySelectorAll('.var-option')).forEach((el) => trigger(el, 'click'))
  expect(wrapper.vm.value).toStrictEqual(['吃饭', '睡觉'])
  await delay(16)

  await wrapper.find('.var-chip--close').trigger('click')
  expect(wrapper.vm.value).toStrictEqual(['睡觉'])

  wrapper.unmount()
})

test('test select validation', async () => {
  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: '',
      }),
      template: `
      <var-select ref="select" :rules="[v => !!v || '您必须选择一个']" v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  const { select } = wrapper.vm.$refs

  select.validate()
  await delay(16)
  expect(wrapper.find('.var-form-details__message').text()).toBe('您必须选择一个')
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-select__wrap').trigger('click')
  await trigger(document.querySelector('.var-option'), 'click')
  await delay(16)
  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()

  select.reset()
  expect(wrapper.vm.value).toBe(undefined)

  wrapper.unmount()
})

test('test select focus & blur methods', async () => {
  const wrapper = mount(
    {
      ...Wrapper,
      data: () => ({
        value: '',
      }),
      template: `
      <var-select ref="select" :rules="[v => !!v || '您必须选择一个']" v-model="value">
        <var-option label="吃饭" />
        <var-option label="睡觉" />
      </var-select>
    `,
    },
    { attachTo: document.body }
  )

  const { select } = wrapper.vm.$refs

  select.focus()
  await delay(16)
  expect(wrapper.html()).toMatchSnapshot()

  select.blur()
  await delay(16)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.unmount()
})

test('test select offset-y', async () => {
  const wrapper = mount(
    {
      components: {
        [VarSelect.name]: VarSelect,
        [VarOption.name]: VarOption,
      },
      data: () => ({
        offsetY: 40,
      }),
      template: `
      <var-select ref="select" offset-y="40">
        <var-option label="火猫" />
        <var-option label="土猫" />
        <var-option label="紫猫" />
        <var-option label="蓝猫" />
      </var-select>
      `,
    },
    { attachTo: document.body }
  )

  const { select } = wrapper.vm.$refs

  select.focus()
  await delay(1000)

  const menu = wrapper.findComponent({ name: 'var-menu-v2' })

  expect(menu.vm.offsetY).toBe(40)

  wrapper.unmount()
})
