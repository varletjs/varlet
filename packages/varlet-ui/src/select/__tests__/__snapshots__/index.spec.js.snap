// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test select example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">朴素模式</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon var-select--non-hint\\"></div>
    <div class=\\"var-select__wrap var-select--non-hint\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis var-select--placeholder-non-hint\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon var-select--non-hint\\">
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">文本关联值</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"relation\\">当前选择的是: </div>
<div class=\\"app-type\\">禁用</div>
<div class=\\"var-select var--box var-select--disabled\\">
  <div class=\\"var-select__controller var-select--disabled\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select var-select--disabled\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis var-select--disabled\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line var-select--line-disabled\\">
    <div class=\\"var-select__dot var-select--line-disabled\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">只读</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">可清除</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\"><i class=\\"var-icon var-icon--set var-icon-close-circle var-select__clear-icon\\" style=\\"transition: transform 0ms; font-size: 14px;\\"></i></div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">自定义图标</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"><i class=\\"var-icon var-icon--set var-icon-plus prepend-icon\\" style=\\"transition: transform 0ms;\\"></i></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-chevron-down\\" style=\\"transition: transform 300ms;\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\"><i class=\\"var-icon var-icon--set var-icon-minus append-icon\\" style=\\"transition: transform 0ms;\\"></i></div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">多选</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\">
          <div>
            <div class=\\"var-select__values\\"></div>
          </div><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i>
        </div><label class=\\"var-select__placeholder var--ellipsis\\">请选择多个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">纸片风格的多选</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\">
          <div>
            <div class=\\"var-select__chips\\"></div>
          </div><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i>
        </div><label class=\\"var-select__placeholder var--ellipsis\\">请选择多个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">单选值校验</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\">请选择一个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"app-type\\">多选值校验</div>
<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\">
          <div>
            <div class=\\"var-select__values\\"></div>
          </div><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i>
        </div><label class=\\"var-select__placeholder var--ellipsis\\">请选择多个选项</label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
<div class=\\"space\\"></div>"
`;

exports[`test select focus & blur methods 1`] = `
"<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller var-select--focus\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow var-select--arrow-rotate\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis var-select--focus var-select--placeholder-hint\\"></label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot var-select--spread\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test select focus & blur methods 2`] = `
"<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis\\"></label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test select hint to be false 1`] = `
"<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller\\">
    <div class=\\"var-select__icon var-select--non-hint\\"></div>
    <div class=\\"var-select__wrap var-select--non-hint\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis var-select--placeholder-hidden var-select--placeholder-non-hint\\"></label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon var-select--non-hint\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line\\">
    <div class=\\"var-select__dot\\"></div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`test select validation 1`] = `
"<div class=\\"var-select var--box\\">
  <div class=\\"var-select__controller var-select--error\\">
    <div class=\\"var-select__icon\\"></div>
    <div class=\\"var-select__wrap\\">
      <div class=\\"var-menu var-select__menu\\" var-select-cover=\\"\\">
        <div class=\\"var-select__select var-select--error\\" style=\\"text-align: left;\\"><span></span><i class=\\"var-icon var-icon--set var-icon-menu-down var-select__arrow\\" style=\\"transition: transform 300ms;\\" var-select-cover=\\"\\"></i></div><label class=\\"var-select__placeholder var--ellipsis var-select--error\\"></label>
        <!--teleport start-->
        <!--teleport end-->
      </div>
    </div>
    <div class=\\"var-select__icon\\">
      <!--v-if-->
    </div>
  </div>
  <div class=\\"var-select__line var-select--line-error\\">
    <div class=\\"var-select__dot var-select--line-error\\"></div>
  </div>
  <transition-stub>
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">您必须选择一个</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
</div>"
`;
